### PR TITLE
AWS for cloud-controller-manager

### LINKED ISSUE #47214: cloudprovider AWS needs implementations for *ByProviderID methods
https://github.com/kubernetes/kubernetes/blob/4220b7303e3ac92e67c9644b4c5bb91621108961/pkg/cloudprovider/providers/aws/aws.go#L1025

https://github.com/kubernetes/kubernetes/blob/4220b7303e3ac92e67c9644b4c5bb91621108961/pkg/cloudprovider/providers/aws/aws.go#L1064

These two methods currently just throw "unimplemented" errors.

### PR DESCRIPTION
**Pull request by ublubu**
fixes #47214

This implements the NodeAddressesByProviderID and InstanceTypeByProviderID methods used by the cloud-controller-manager for the AWS provider.

NodeAddressesByProvider uses DescribeInstances (for normal addresses) and DescribeAddresses (for Elastic IP addresses).

InstanceTypeByProviderID uses DescribeInstances.

```release-note
NONE
```

### CODE REVIEW

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -3419,6 +3518,78 @@ func (c *Cloud) getFullInstance(nodeName types.NodeName) (*awsInstance, *ec2.Ins
...
+
+var providerIDRegexp = regexp.MustCompile(`^aws://([^/]+)$`)
+
+func instanceIDFromProviderID(providerID string) (instanceID string, err error) {
+	matches := providerIDRegexp.FindStringSubmatch(providerID)
+	if len(matches) != 1 {
+		return "", fmt.Errorf("ProviderID \"%s\" didn't match expected format \"aws://InstanceID\"", providerID)
+	}
+
+	return matches[1], nil
```
Comment (wlan0) [inline pkg/cloudprovider/providers/aws/aws.go]: This will cause OutOfIndex error.

**Comment by wlan0** [comment]
@ublubu how was the kubelet obtaining node addresses before? Was it using the metadata service? 

I'm asking because I'm wondering why there is a new function to describe addresses. Shouldn't there be existing methods for this very purpose?

**Comment by ublubu** [comment]
@wlan0 The kubelet uses the local metadata service.

In the kubelet:
https://github.com/kubernetes/kubernetes/blob/b8c9ee8abb387de7d6e0e910538bd53057c77d0f/pkg/kubelet/kubelet_node_status.go#L432

AWS implementation of NodeAddresses using metadata service:
https://github.com/kubernetes/kubernetes/blob/b8c9ee8abb387de7d6e0e910538bd53057c77d0f/pkg/cloudprovider/providers/aws/aws.go#L940

IIUC, the new methods are run in the CCM on a master, where a node's metadata service is not available.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (ublubu) [inline pkg/cloudprovider/providers/aws/aws.go]: Oh whoops. Fixed it and added a test for this.

**Comment by wlan0** [comment]
@ublubu The code looks good. Thanks for the PR!

LGTM

@thockin @luxas PTAL

**Comment by thockin** [comment]
@k8s-bot ok to test

**Comment by thockin** [comment]
/lgtm
/approve

@dchen1107 same question - 1.7.0 or 1.7.1?

**Comment by wlan0** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by wlan0** [comment]
Addresses https://github.com/kubernetes/kubernetes/issues/47257

**Comment by thockin** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by luxas** [comment]
@thockin May I set the v1.7 label on this?
The risk (if any) seems to be really low...

**Comment by thockin** [comment]
@dchen1107 this is thoroughly flag gated already

**Comment by justinsb** [comment]
It seems this PR uses an incompatible provider id format.

**Comment by wlan0** [comment]
What should the format be?

**Comment by justinsb** [comment]
Addressed in #47395 I believe

**Comment by wlan0** [comment]
@justinsb Thanks, Will fix it now

**Pull request merged**